### PR TITLE
Added possibility to load fixtures by bundle name

### PR DESCRIPTION
**Pull request by biozshock**
Feature allows loading fixtures from bundles.

``` bash
app/console doctrine:fixtures:load --fixtures=@AcmeDemoBundle
```

will load `/path/to/src/Acme/DemoBundle/DataFixtures/ORM`

### CODE REVIEW

**Comment by lavoiesl** [comment]
Could you rebase your PR?

**Comment by biozshock** [comment]
Huh.. Almost 3 years :) On top of which branch to rebase? I guess 2.0 so more people can use this now?

**Comment by stof** [comment]
On top of master. The 2.0 branch is an abandonned branch for Symfony 2.0 (which is long dead)

**Comment by biozshock** [comment]
Huh. Thought that are branches for Symfony 2 and 3.
Rebased.

**Comment by lavoiesl** [comment]
@biozshock, sorry, would it be possible to add a test for this as well?

File: composer.json [hunk 1]
```diff
@@ -22,9 +22,14 @@
     "require": {
         "php": ">=5.3.2",
         "symfony/doctrine-bridge": "~2.1",
+        "symfony/console": "~2.1",
         "doctrine/doctrine-bundle": "~1.0",
+        "doctrine/orm": "~2.0",
```
Comment (lavoiesl) [inline composer.json]: This bundle should not depend on ORM

File: Tests/bootstrap.php [hunk 2]
```diff
@@ -0,0 +1,3 @@
+<?php
+
+$loader = require_once(__DIR__ . '/../vendor/autoload.php');
```
Comment (lavoiesl) [inline Tests/bootstrap.php]: Missing newline

File: Tests/Fixtures/Entity/CheckIt.php [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    /**
+     * @param string $email
+     */
+    public function setEmail($email)
+    {
+        $this->email = $email;
+    }
+
+
+}
```
Comment (lavoiesl) [inline Tests/Fixtures/Entity/CheckIt.php]: Missing newline

File: Tests/Fixtures/DataFixtures/ORM/CheckLoad.php [hunk 4]
```diff
@@ -0,0 +1,28 @@
+<?php
+/**
+ * Created by PhpStorm.
```
Comment (lavoiesl) [inline Tests/Fixtures/DataFixtures/ORM/CheckLoad.php]: Wrong copyright header, remove it or add the standard Doctrine header.

File: Tests/Fixtures/DataFixtures/ORM/CheckLoad2.php [hunk 5]
```diff
@@ -0,0 +1,28 @@
+<?php
+/**
+ * Created by PhpStorm.
```
Comment (lavoiesl) [inline Tests/Fixtures/DataFixtures/ORM/CheckLoad2.php]: Wrong copyright header, remove it or add the standard Doctrine header.

File: Tests/Fixtures/DataFixtures/ORM/Product/CheckLoad3.php [hunk 6]
```diff
@@ -0,0 +1,28 @@
+<?php
+/**
+ * Created by PhpStorm.
```
Comment (lavoiesl) [inline Tests/Fixtures/DataFixtures/ORM/Product/CheckLoad3.php]: Wrong copyright header, remove it or add the standard Doctrine header.

File: Tests/Fixtures/Entity/CheckIt.php [hunk 7]
```diff
@@ -0,0 +1,66 @@
+<?php
+/**
+ * Created by PhpStorm.
```
Comment (lavoiesl) [inline Tests/Fixtures/Entity/CheckIt.php]: Wrong copyright header, remove it or add the standard Doctrine header.

File: Tests/Fixtures/DataFixtures/ORM/CheckLoad2.php [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+     */
+    public function load(ObjectManager $manager)
+    {
+        $userAdmin = new CheckIt();
+        $userAdmin->setEmail('CheckLoad2@got.me');
+
+        $manager->persist($userAdmin);
+        $manager->flush();
+    }
+}
```
Comment (lavoiesl) [inline Tests/Fixtures/DataFixtures/ORM/CheckLoad2.php]: Missing newline

File: Tests/Fixtures/DataFixtures/ORM/CheckLoad.php [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+     */
+    public function load(ObjectManager $manager)
+    {
+        $userAdmin = new CheckIt();
+        $userAdmin->setEmail('CheckLoad@got.me');
+
+        $manager->persist($userAdmin);
+        $manager->flush();
+    }
+}
```
Comment (lavoiesl) [inline Tests/Fixtures/DataFixtures/ORM/CheckLoad.php]: Missing newline

File: Tests/Command/LoadDataFixturesDoctrineCommandTest.php [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+        $bundle = $this->getMock('Symfony\Component\HttpKernel\Bundle\BundleInterface');
+        $bundle
+            ->expects($this->any())
+            ->method('getPath')
+            ->will($this->returnValue(realpath(__DIR__ . $path)))
+        ;
+
+        return $bundle;
+    }
+}
```
Comment (lavoiesl) [inline Tests/Command/LoadDataFixturesDoctrineCommandTest.php]: Missing newline

File: Tests/Command/LoadDataFixturesDoctrineCommandTest.php [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+            ->will($this->returnValue(array($this->getBundle())))
+        ;
+
+        return $this->kernel;
+    }
+
+    private function getBundle($path = '/../Fixtures')
+    {
+        $bundle = $this->getMock('Symfony\Component\HttpKernel\Bundle\BundleInterface');
+        $bundle
```
Comment (lavoiesl) [inline Tests/Command/LoadDataFixturesDoctrineCommandTest.php]: Expected formatting is:
```php
$bundle->expects($this->any())
    ->method('getPath')
    ->will($this->returnValue(realpath(__DIR__ . $path)));
```

File: Tests/Command/LoadDataFixturesDoctrineCommandTest.php [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+
+    private function getMockKernel()
+    {
+        $this->kernel = $this->getMock('Symfony\Component\HttpKernel\KernelInterface');
+        $this->kernel
+            ->expects($this->any())
+            ->method('getBundle')
+            ->will($this->returnValue($this->getBundle()))
+        ;
+        $this->kernel
```
Comment (lavoiesl) [inline Tests/Command/LoadDataFixturesDoctrineCommandTest.php]: formatting, see below

File: Tests/Command/LoadDataFixturesDoctrineCommandTest.php [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+                call_user_func($callable, $em);
+            }));
+
+        return $container;
+    }
+
+    private function getMockKernel()
+    {
+        $this->kernel = $this->getMock('Symfony\Component\HttpKernel\KernelInterface');
+        $this->kernel
```
Comment (lavoiesl) [inline Tests/Command/LoadDataFixturesDoctrineCommandTest.php]: formatting, see below

File: Tests/Command/LoadDataFixturesDoctrineCommandTest.php [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+        );
+    }
+
+    /**
+     * @depends testFixtureLoad
+     * @expectedException \InvalidArgumentException
+     */
+    public function testNoFixtures()
+    {
+        $this->testFixtureLoad('hoho', array());
```
Comment (lavoiesl) [inline Tests/Command/LoadDataFixturesDoctrineCommandTest.php]: use common strings like `'nonexistant'` or `'foo'`

File: Command/LoadDataFixturesDoctrineCommand.php [hunk 15]
```diff
@@ -75,13 +79,25 @@ protected function execute(InputInterface $input, OutputInterface $output)
...
+        $fixturesPath = DIRECTORY_SEPARATOR . 'DataFixtures' . DIRECTORY_SEPARATOR . 'ORM';
         $dirOrFile = $input->getOption('fixtures');
         if ($dirOrFile) {
-            $paths = is_array($dirOrFile) ? $dirOrFile : array($dirOrFile);
+
+            $paths = array();
+            $pathsParams = is_array($dirOrFile) ? $dirOrFile : array($dirOrFile);
+            foreach ($pathsParams as $path) {
+                if (preg_match('/^\@([^\/\\\\]+)(.*)$/', $path, $matches)) {
+                    $bundle = $this->getApplication()->getKernel()->getBundle($matches[1]);
```
Comment (lavoiesl) [inline Command/LoadDataFixturesDoctrineCommand.php]: When the bundle does not exist (or typo), is it handled properly? There should be a test case for that.

File: composer.json [hunk 16]
```diff
@@ -22,9 +22,14 @@
     "require": {
         "php": ">=5.3.2",
         "symfony/doctrine-bridge": "~2.1",
+        "symfony/console": "~2.1",
```
Comment (lavoiesl) [inline composer.json]: This bundle should not depend on Console, add it as require-dev if you must, like in the library: https://github.com/doctrine/data-fixtures/blob/master/composer.json.

File: Command/LoadDataFixturesDoctrineCommand.php (hunk 15 above)
Comment (biozshock) [inline Command/LoadDataFixturesDoctrineCommand.php]: Out of scope of responsibility for this code. This should be handled by Kernel. But will add a testcase which mimic this behaviour.

File: composer.json (hunk 1 above)
Comment (biozshock) [inline composer.json]: FYI: This code will not work without ORM as it has a hard dependency on Doctrine\ORM\EntityManagerInterface in \Doctrine\Common\DataFixtures\Purger\ORMPurger.
But this is up to you :)

File: Command/LoadDataFixturesDoctrineCommand.php (hunk 15 above)
Comment (lavoiesl) [inline Command/LoadDataFixturesDoctrineCommand.php]: Excellent, this is exactly what I was looking for.

File: Command/LoadDataFixturesDoctrineCommand.php [hunk 17]
```diff
@@ -100,7 +115,7 @@ protected function execute(InputInterface $input, OutputInterface $output)
         $purger = new ORMPurger($em);
         $purger->setPurgeMode($input->getOption('purge-with-truncate') ? ORMPurger::PURGE_MODE_TRUNCATE : ORMPurger::PURGE_MODE_DELETE);
         $executor = new ORMExecutor($em, $purger);
-        $executor->setLogger(function($message) use ($output) {
+        $executor->setLogger(function ($message) use ($output) {
```
Comment (lavoiesl) [inline Command/LoadDataFixturesDoctrineCommand.php]: Sorry to be nitpick, but this shouldn’t be modified.

File: Tests/Command/LoadDataFixturesDoctrineCommandTest.php [hunk 18]
```diff
@@ -0,0 +1,149 @@
...
+            ->method('get')
+            ->with('doctrine')
+            ->will($this->returnValue($doctrineMock));
+
+        $this->emMock = $this->getMock('Doctrine\ORM\EntityManagerInterface');
+        $doctrineMock->expects($this->any())
+            ->method('getManager')
+            ->will($this->returnValue($this->emMock));
+
+        $doctrineEventMock = $this->getMock('\Doctrine\Common\EventManager');
```
Comment (lavoiesl) [inline Tests/Command/LoadDataFixturesDoctrineCommandTest.php]: For consistency, please remove leading slash

File: Command/LoadDataFixturesDoctrineCommand.php (hunk 17 above)
Comment (biozshock) [inline Command/LoadDataFixturesDoctrineCommand.php]: phpcs reported this line. https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-2-coding-style-guide.md#6-closures

Comment (lavoiesl) [inline Command/LoadDataFixturesDoctrineCommand.php]: I understand, but this is not the job of this PR.

File: composer.json (hunk 1 above)
Comment (lavoiesl) [inline composer.json]: This bundle was working before your modification and your modification is not adding a specific dependency on ORMPurger. Thus, I would advise not to modify this. If this is to be changed, it would need to be in a separate PR.

**Comment by biozshock** [comment]
@lavoiesl I think it's fine to merge this now :)

**Comment by alcaeus** [comment]
#209 changes the fixture loader to consider tagged services, making bundle integration easier. Closing here.
